pysoa/common/transport/redis_gateway/core.py

```diff
@@ -121,14 +127,19 @@ def __attrs_post_init__(self):
     @property
     def backend_layer(self):
         if self._backend_layer is None:
-            backend_layer_kwargs = deepcopy(self.backend_layer_kwargs)
-            with self._get_timer('backend.initialize'):
-                if self.backend_type == REDIS_BACKEND_TYPE_SENTINEL:
-                    self._backend_layer = SentinelRedisClient(**backend_layer_kwargs)
-                    self._backend_layer.metrics_counter_getter = lambda name: self._get_counter(name)
-                else:
-                    self._backend_layer = StandardRedisClient(**backend_layer_kwargs)
-
+            cache_key = repr((self.backend_type, self.backend_layer_kwargs))
```

Comment:
nitpick: `self.backend_layer_kwargs` is a dict, it will probably work fine because even in Py2 within thread repr is consistent, but do we want to rely on undocumented behavior?

Reply:
For two or more dicts with the same content, `repr(each_dict)` is guaranteed to be the same within a given Python process. It's only guaranteed to be the same across processes in Python 3, but in Python 2 it's guaranteed to be the same within the same process, which is all we need.

Reply:
I turned out to be incorrect here. Fix incoming.

Reply:
For reference:

```
>>> d1 = {u'hosts': [u'1.2.3.4', u'5.6.7.8', u'9.10.11.12'], u'connection_kwargs': {u'password': u'abc123meta', u'socket_connect_timeout': 20}, u'sentinel_services': [u'service1', u'service2'], u'sentinel_failover_retries': 6}
>>> d2 = {u'sentinel_services': [u'service1', u'service2'], u'sentinel_failover_retries': 6, u'connection_kwargs': {u'password': u'abc123meta', u'socket_connect_timeout': 20}, u'hosts': [u'1.2.3.4', u'5.6.7.8', u'9.10.11.12']}
>>> d3 = {u'connection_kwargs': {u'socket_connect_timeout': 20, u'password': u'abc123meta'}, u'sentinel_services': [u'service1', u'service2'], u'sentinel_failover_retries': 6, u'hosts': [u'1.2.3.4', u'5.6.7.8', u'9.10.11.12']}
>>> d1 == d2
True
>>> d1 == d3
True
>>> repr(d1) == repr(d2)
False
>>> repr(d1) == repr(d3)
True
```

But:

```
>>> %timeit repr(d1)
The slowest run took 5.76 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 3.48 µs per loop

>>> %timeit dict_to_hashable(d1)
The slowest run took 4.16 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 4.07 µs per loop
```

And the result of `dict_to_hashable` can be used directly as a dict key.